formats/mdm/MDM_2_4.ddl

```diff
@@ -0,0 +1,515 @@
+    -- VRT_Packet_Payload -- M
+    -- VRT_Trailer -- Packet dependent
+
+    ^ {| Body_VRT = { } |}
+
+-- See 6.3.1.2.2: VRT Prologue
+def VRT_Prologue =
+  block
+    header = VRT_Packet_Header
+    stream_identifier = VRT_Stream_Identifier
+
+    -- The Class Identifier is present if and only if the "C" bit of the
+    -- header is set.
+    class_id = if header.c_bit == 1
+                 then just VRT_Class_Identifier
+                 else nothing
+
+    -- Each timestamp field is present if and only if the appropriate
+    -- header indicator is present.
+    timestamp_int = case header.tsi of
+                      No_Int_Field -> nothing
+                      _ -> just BEUInt32
+
+    timestamp_frac = case header.tsf of
+                       No_Frac_Field -> nothing
+                       _ -> just BEUInt64
+
+    -- TODO: use:
+    -- VRT_Packet_Body header
+
+-- Parse the VRT packet body, including the packet-specific prologue,
+-- payload, and trailer, as appropriate to the packet type.
+def VRT_Packet_Body (header: VRT_Packet_Header) =
+  block
+    -- TODO: the MORA spec elides many details from the Vita 49.*
+    -- specification(s) that are needed to fully parse the message body
+    -- and trailer.
+    Fail "VRT packet body parsing not implemented"
+
+def VRT_Packet_Header =
+  BEUInt32 as? VRT_Packet_Header_bits
+
+def VRT_Class_Identifier =
+  BEUInt64 as? VRT_Class_Identifier_bits
+
+bitdata VRT_Information_Class_Code_ID where
+  M = 0x4d: uint 8
+
+bitdata VRT_Information_Class_Code where
+  id: VRT_Information_Class_Code_ID
+  version_maj: uint 4
+  version_min: uint 4
+
+bitdata VRT_Packet_Class_Code where
+  reserved: uint 12
+  packet_type: VRT_Packet_Type
+
+bitdata VRT_Class_Identifier_bits where
+  pad_bit_count: uint 5
+  reserved: uint 3
+  org_unique_identifier: uint 24
+  information_class_code: VRT_Information_Class_Code
+  packet_class_code: VRT_Packet_Class_Code
+
+-- See 6.3.1.2.4: VRT Stream ID
+def VRT_Stream_Identifier =
+  block
+    resource_id = BEUInt16 as? Identifier_Triple
+    subport_id = BEUInt16
+
+-- See 6.3.1.2.3: VRT Packet Header Field
+bitdata VRT_Packet_Header_bits where
+  type: VRT_Packet_Type
+  c_bit: uint 1
+  indicators: uint 3
+  tsi: Timestamp_Int_Status
+  tsf: Timestamp_Frac_Status
+  packet_count: uint 4
+  packet_size_words: uint 16
+
+-- See 6.3.1.2.3: VRT Packet Header Field
+bitdata Timestamp_Int_Status where
+  No_Int_Field = 0x0: uint 2
+  UTC_Relative = 0x1: uint 2
+
+-- See 6.3.1.2.3: VRT Packet Header Field
+bitdata Timestamp_Frac_Status where
+  No_Frac_Field         = 0x0: uint 2
+  Real_Time_Picoseconds = 0x2: uint 2
+
+-- See 6.3.1.2.3: VRT Packet Header Field
+bitdata VRT_Packet_Type where
+  Signal_Data_Packet_w_Stream_ID = 0x1: uint 4
+  Context_Packet                 = 0x4: uint 4
+  Command_Packet                 = 0x6: uint 4
+
+-- See 6.3.7: MDM Switch Group User ID Message (Type 7) Specification
+def Parse_Body_Switch_Group_User_ID: MDM_Body_u =
```

Comment:
You don't really need all the `let`s here, you could just inline them.

Also, it might be nicer to just factor out the payload into its own parser.